### PR TITLE
refactor: make it easier to reason about startup exit

### PR DESCRIPTION
**Pull request by toidiu**
This PR refactors the startup on_congestion code to more clearly identify why we exit the startup phase. We will use this in a follow up PR to emit metrics about exit reason.

### CODE REVIEW

File: quiche/src/recovery/gcongestion/bbr2/startup.rs [hunk 1]
```diff
@@ -64,26 +65,42 @@ impl ModeImpl for Startup {
...
+        let mut exit_due_to_excessive_loss = false;
+
         let has_bandwidth_growth =
             self.model.has_bandwidth_growth(congestion_event, params);
 
-        #[allow(clippy::absurd_extreme_comparisons)]
-        if params.max_startup_queue_rounds > 0 && !has_bandwidth_growth {
+        let check_bw_plateau =
+            params.max_startup_queue_rounds > 0 && !has_bandwidth_growth;
+        if check_bw_plateau {
```
[the remark this answers is not in the document]
Comment (toidiu) [reply quiche/src/recovery/gcongestion/bbr2/startup.rs]: I was pulling out a comparison into a variable but based on your comment it doesn't seem worth it.

The `absurd_` lint was needed when we were using the [PARAM](https://github.com/cloudflare/quiche/blob/f6a1927f3c7acb79b2f9c7110c5f753475437bec/quiche/src/recovery/gcongestion/bbr2/startup.rs#L70) global value, which was set to `0`.

File: quiche/src/recovery/gcongestion/bbr2/probe_bw.rs [hunk 2]
```diff
@@ -394,11 +395,14 @@ impl ProbeBW {
         } else if self.cycle.rounds_in_phase > 0 {
             if params.max_probe_up_queue_rounds > 0 {
                 if congestion_event.end_of_round_trip {
-                    self.model
+                    let PersistentQueueOutcome {
+                        full_bandwidth_reached: _,
+                        rounds_with_queueing,
+                    } = self
```
[the remark this answers is not in the document]
Comment (toidiu) [reply quiche/src/recovery/gcongestion/bbr2/probe_bw.rs:400]: It better for type safety to avoid the `..` here and force the caller to explicitly not use a value.

We have a lot of side-effecty functions which makes it difficult to know whats changing and what needs to be checked. You caught this in the [comment below](https://github.com/cloudflare/quiche/pull/2061/files?diff=split&w=0#r2096591276).

I annotated `PersistentQueueOutcome` with `#[must_use]`, which tells the compiler to enforce usage (the function had a side_effect so make sure the caller of the function inspects the side_effect).

By using `..` we nullify the benefit of `must_use` if we add a new value to PersistentQueueOutcome in the future.

```
PersistentQueueOutcome {
   rounds_with_queueing,
  new_side_effect // THIS will automatically be ignored without a warning
}
```

File: quiche/src/recovery/gcongestion/bbr2/startup.rs [hunk 3]
```diff
@@ -64,26 +65,42 @@ impl ModeImpl for Startup {
...
-        if !congestion_event.last_packet_send_state.is_app_limited &&
-            !has_bandwidth_growth
-        {
-            self.check_excessive_losses(congestion_event, params);
+            let PersistentQueueOutcome {
+                full_bandwidth_reached,
+                rounds_with_queueing: _,
+            } = self.model.check_persistent_queue(1.75, params);
+
+            exit_due_to_bw_plateau = full_bandwidth_reached;
```
[the remark this answers is not in the document]
Comment (toidiu) [reply quiche/src/recovery/gcongestion/bbr2/startup.rs]: Good catch!

File: quiche/src/recovery/gcongestion/bbr2/startup.rs [hunk 4]
```diff
@@ -64,26 +65,42 @@ impl ModeImpl for Startup {
...
+            params.max_startup_queue_rounds > 0 && !has_bandwidth_growth;
+        if check_bw_plateau {
             // 1.75 is less than the 2x CWND gain, but substantially more than
             // 1.25x, the minimum bandwidth increase expected during
             // STARTUP.
-            self.model.check_persistent_queue(1.75, params);
-        }
-        // TCP BBR always exits upon excessive losses. QUIC BBRv1 does not exit
-        // upon excessive losses, if enough bandwidth growth is observed or if the
-        // sample was app limited.
```
[the remark this answers is not in the document]
Comment (toidiu) [reply quiche/src/recovery/gcongestion/bbr2/startup.rs:79]: There seem to be a lot of assumptions baked in. I read this as saying the following and its definitely possible for these assumption to not be true:
- we don't believe our startup behavior is "aggressive" enough to induce significant loss when app limited
- if we are app limited, then loss was not due to congestion that we introduced so continue in startup

[the remark this answers is not in the document]
Comment (antoniovicente) [reply quiche/src/recovery/gcongestion/bbr2/startup.rs:79]: It is possible to send more than the previous round and still be app limited.  BW growth when minRTT remains constant implies that we managed to send and ack more data than the previous round.

I have a lot of worries about app limited being computed correctly since it seems to be relatively magic in nature.  Or the issue may be that I haven't learned enough about the definition of app limited yet.

File: quiche/src/recovery/gcongestion/bbr2/startup.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (antoniovicente) [reply quiche/src/recovery/gcongestion/bbr2/startup.rs]: Good to know that the clippy lint is out of date.

Introducing a local variable as documentation generally makes sense to me.  So by all means, keep this kind of change.

File: quiche/src/recovery/gcongestion/bbr2/startup.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (toidiu) [reply quiche/src/recovery/gcongestion/bbr2/startup.rs:79]: These are good questions ^

File: quiche/src/recovery/gcongestion/bbr2/probe_bw.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (antoniovicente) [reply quiche/src/recovery/gcongestion/bbr2/probe_bw.rs:401]: Thanks.  You make a compelling argument.

File: quiche/src/recovery/gcongestion/bbr2/startup.rs [hunk 5]
```diff
@@ -66,21 +67,36 @@ impl ModeImpl for Startup {
...
             // STARTUP.
-            self.model.check_persistent_queue(1.75, params);
-        }
+            let PersistentQueueOutcome {
+                full_bandwidth_reached,
+                rounds_with_queueing: _,
+            } = self.model.check_persistent_queue(1.75, params);
+
+            // TODO: Exit due to persistent queue
+            let _exit_due_to_persistent_queue = full_bandwidth_reached;
```
Comment (antoniovicente) [inline quiche/src/recovery/gcongestion/bbr2/startup.rs:85]: The approach in https://github.com/cloudflare/quiche/pull/2057 has evolved a bit.  I would consider not exposing full_bandwidth_reached in PersistentQueueOutcome yet.  And remove the TODOs associated with 2057.

It would be good to have a version of 2057 rebased to take advantage of changes in this PR.

**Comment by toidiu** [comment]
replaced by https://github.com/cloudflare/quiche/pull/2064
